File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #input - array nums,output-array of nums
        #brute - 2 loops find product for one element for each of the other elements
        #optimal - find prefix and suffix product for each element and multiply them

        nums_length = len(nums)

        prefix = [1]*nums_length
        suffix = [1]*nums_length

        res=[]

        for i in range(1,nums_length):
            prefix[i] = prefix[i-1]*nums[i-1]
        
        for i in range(nums_length-2,-1,-1):
            suffix[i] = suffix[i+1]*nums[i+1]

        for i in range(nums_length):
            res.append(prefix[i] * suffix[i])

        return res



        

File: Data Structures & Algorithms/duplicate-integer/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {boolean}
     */
    hasDuplicate(nums) {
        const map = new Map();
        for (var i=0;i<nums.length;i++){
            if (map.has(nums[i])){ 
                map.set(nums[i],map.get(nums[i])+1);
            }
            else{
                map.set(nums[i],1);
            }
        }
        console.log(map)
        const res=0;
        for (let [key,val] of map){
            if(val>1){
                return true;
            }
        }
        return false;
    }
}
